### ISSUE TITLE
Ability to translate a group of translation keys at once

### ISSUE THREAD

**Issue by binokaryg** [medic/medic#5969]
**What feature do you want to improve?**
It would be better if we could translate a group of keys at once so that they appear together, separated by a delimiter.

**Describe the improvement you'd like**
I have an array of risk factors which may vary from person to person. These risk factors need to be shown together (separated by semicolon). I want to be able to translate them. It would be nice to have a filter that accepts an array of translation keys and shows it in translated form.
It also applies to the danger signs.

**Describe alternatives you've considered**
It might be possible to show the risk factors in a different language by reading user locale during the contact-summary generation, but for that, we need to keep the translations for all supported languages inside the contact-summary logic, which is probably not a good idea.

**Additional context**
Here is how it looks with current implementation in English:
![image](https://user-images.githubusercontent.com/20041732/65302451-f26a2580-db9a-11e9-88cc-794c1c9021c4.png)

**Comment by binokaryg** [comment]
@abbyad and I just tried nested translations (#4599):

First try:
```
danger.first = first pregnancy
danger.highpressure = high blood pressure
```
```
 { label: 'contact.profile.danger_signs.current', value: '${danger.first}, ${danger.highpressure}', translate: true, width: 6 }
```
It didn't work at first.
After setting the translations in a different way:
```
danger.first = first pregnancy
danger.highpressure = high blood pressure
danger.multiple = ${danger.first}, ${danger.highpressure}
```
```
 { label: 'contact.profile.danger_signs.current', value: 'danger.multiple', translate: true, width: 6 }
```
It was possible to show comma separated values, but this doesn't solve the problem, as we can't dynamically select the danger signs to show in the contact-summary this way.

**Comment by abbyad** [comment]
Too bad that the nested translations didn't work in this case! 

Being able to use nested translations here could be useful, but it might be better to show this information as an actual list. I have confirmed with @ranjuts that this would be preferable in this scenario since the content would be easier to read. 

To show a list of items we could structure the list items in `value` as an array. Here is an example using the same scenario from:

` { label: 'contact.profile.danger_signs.current', value: ['danger.first', 'danger.highpressure'], translate: true, width: 6 }`

**Comment by SCdF** [comment]
@binokaryg / @abbyad I have prototyped this, and with the following config example:
```js
{ appliesToType: 'person', label: 'Stefan Test', value: ['contact', 'contact.phone', 'contact.notes'], translate: true, width: 12 }
```

I get this:
![image](https://user-images.githubusercontent.com/583851/73378197-21ff0c00-42b8-11ea-8134-8e60242ff44b.png)

Is this what you're looking for? If so I will polish and generate a PR.

**Comment by binokaryg** [comment]
This looks good enough to me @SCdF.
Any possibility that a different delimiter can be shown in the output, such as semicolons instead of commas? Shall we make it customizable?

**Comment by SCdF** [comment]
Is it deeply important to be able to do this? If the goal is that we want to output translatable lists then I’d rather we keep it simple and just define it once. If certain cultures use different delimiters for lists we can look at supporting that automatically

**Comment by SCdF** [comment]
Looking deeper into this, I want to solidify how this change interacts with the existing API.

`filter` and `context` are the two that I'm aware of that work on or with the `value` parameter.

My goal here is to make sure we are not creating unintuitive combinations of the existing feature set, not to create new features that we don't know we need.
- `filter`: seems logical to me this stays as a single function, and if you provide `filter` with a `value[]` it applies that function to each element in the array. E.g., you can combine an ordered array of DOBs with a filter function to list the ages of children, in order.
- `context`: more challenging I think. Context maps naturally to a single value, and so having an array of them running over the same context seems weird. We'd either need to support `context` being the same structure and length as `value` (singular object, or an array of equal length) or not support it at all. For simplicity I'd suggest for now not supporting it at all, since I don't feel this situation will arise much.

@binokaryg thoughts?

**Comment by binokaryg** [comment]
To select a proper delimiter, we have to consider the possible usage.
For example, let's suppose we need to display the following list of immunization schedule:

- Chickenpox
- Diphtheria, Tetanus, and Whooping cough
- Haemophilus Influenzae type B
- Measles, Mumps, Rubella

Separating the above items with commas may not correctly express the fact that there are actually 4 items here.

Also, I like the `filter`/`value` combination.
Using `filter`/`values` is also fine if it can actually reduce the risk of confusion.

**Comment by abbyad** [comment]
Danger signs like in the original post can be the result of a multi-select field. Could we handle that situation without having to manipulate it heavily on the JS side?

**Comment by SCdF** [comment]
@abbyad could you go a bit deeper on your example here. The picture included in this ticket description should be achievable with the design I'm working on.

**Comment by abbyad** [comment]
I'll put together an example XForm that you could test with.

**Comment by binokaryg** [comment]
In the document, the items of a multi-select field are separated by space.
e.g. `"danger_signs": "first high_pressure multiple"`
It should be easy to split them in javascript. However, the translations need to match the items exactly.
If the translations have some prefix, similar to those in the original post (`danger.`), some manipulation will be necessary:
```
var ds_field = "first high_pressure multiple";
var danger_signs = ds_field.split(' ').map(x => 'danger.' + x);
//danger_signs = ["danger.first", "danger.high_pressure", "danger.multiple"]
```

**Comment by SCdF** [comment]
Right OK I think in those situations we are going to have to rely on JS being used. 

I would suggest that if we find that XForms always generates things like this and it's a constant annoyance, we could add a function into our JS context which can be used to deal with this.

However, I wouldn't want to bake it into the supported output format, because that reduces our flexibility by a large amount (ie if you record data **not** this way)

**Comment by garethbowen** [comment]
@SCdF It looks like there has been no action on this for a while so I'm going to postpone it to help get 3.9.0 out sooner.

**Comment by SCdF** [comment]
@garethbowen yep good plan, the 2-3 things against me had to stop for hiring work, and I won't be able to spin back up on all three in time for 3.9

**Comment by abbyad** [comment]
Just noticing that I never put an example XForm here. We can add one when we come back to this, and in the meantime it's probably enough to know that all multi select XForm fields are saved in the same format, regardless of locale.

**Comment by garethbowen** [comment]
Moved to 3.12.0 to free up engineers to work on 3.11.0.

**Comment by binokaryg** [comment]
Could #10678 address this?
